js/src/common/states/PaginatedListState.ts

```diff
@@ -0,0 +1,171 @@
+import * as Mithril from 'mithril';
+
+import Model from '../Model';
+
+export interface Page<TModel> {
+  number: number;
+  items: TModel[];
+
+  // // maybe we can put page links here?
+  // links: {[name: string]: string},
+  // // or
+  // previousPage?: string;
+  // nextPage?: string;
+}
+
+export interface PaginationLocation {
+  page: number;
+  startIndex?: number;
+  endIndex?: number;
+}
+
+export default abstract class PaginatedListState<T extends Model> {
+  // TODO:
+  //  - what does this page track? last page request? last page that is loaded
+  //      (e.g. pg 7 if u have 3-7 loaded)
+  //  - When do we set it?
+  protected location!: PaginationLocation;
+
+  // TODO: might want interface so pagination can work w/o model-specific stuff
+  protected type: string;
+  protected offset: number;
+
+  protected pages: Page<T>[] = [];
+  protected params: any = {};
+
+  protected initialLoading: boolean = false;
+  protected loadingPrev: boolean = false;
+  protected loadingNext: boolean = false;
+
+  hasPrev: boolean = false;
+  hasNext: boolean = false;
+
+  // TODO: this constructor is for class that implements
+  // what data do we want passed?
+  protected constructor(type: string, offset: number = 20) {
+    this.type = type;
+    this.offset = offset;
+  }
+
+  protected parseResults(pageNum, results: T[]) {
+    const page = {
+      number: pageNum,
+      items: results,
+    };
+
+    // TODO: is this how we were thinking of treating this.pages ?
+    if (this.isEmpty() || pageNum > this.getNextPageNumber() - 1) {
+      this.pages.push(page);
+    } else {
+      this.pages.unshift(page);
```

Comment:
If location.page is a number it'd need to be incremented by 1 when unshifting, unless the page being added is the first page.

Reply:
What would we need to do this for? Sorry, can't follow the logic right now, could you explain it?

Reply:
Just saying we need to update the location page number when pages are added to the start.

Reply:
Ah I see, makes sense. We'd be updating the location page number to the new pages that have been added at the beginning, right?

Reply:
Yep!